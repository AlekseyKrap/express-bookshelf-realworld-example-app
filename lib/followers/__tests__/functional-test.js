'use strict';

const chance = require('chance').Chance('follow');
const request = require('supertest');
const app = require('../../app');

const {followers, users} = app.locals.services;

describe('followers', () => {
  let user;
  let follower;

  beforeEach(async () => {
    user = await users.create({
      bio: chance.profession(),
      email: chance.email(),
      image: chance.avatar(),
      password: chance.word({length: 8}),
      username: chance.word(),
    });
    follower = await users.create({
      bio: chance.profession(),
      email: chance.email(),
      image: chance.avatar(),
      password: chance.word({length: 8}),
      username: chance.word(),
    });
    await followers.create({
      user: user.id,
      follower: follower.id,
    });
  });

  afterEach(async () => {
    await user.destroy();
    await follower.destroy();
  });

  describe('POST /api/profiles/:username/follow', () => {
    describe('without a logged in user', () => {
      test('should return a 401', async () => {
        const {body, status} = await request(app)
          .post(`/api/profiles/${follower.get('username')}/follow`)
          .set('Accept', 'application/json')
          .send();

        expect({body, status}).toMatchSnapshot();
      });
    });

    describe('with a logged in user', () => {
      describe('and an invalid username', () => {
        test('should return a 404', async () => {
          const {body, status} = await request(app)
            .post(`/api/profiles/invalid/follow`)
            .set('Accept', 'application/json')
            .set('Authorization', `Token ${users.generateJWT(user)}`)
            .send();

          expect({body, status}).toMatchSnapshot();
        });
      });

      describe('and a valid username', () => {
        test('should return a 200', async () => {
          const {body, status} = await request(app)
            .post(`/api/profiles/${follower.get('username')}/follow`)
            .set('Accept', 'application/json')
            .set('Authorization', `Token ${users.generateJWT(user)}`)
            .send();

          expect({body, status}).toMatchSnapshot();
        });
      });
    });
  });

  describe('DELETE /api/profiles/:username/follow', () => {
    describe('without a logged in user', () => {
      test('should return a 401', async () => {
        const {body, status} = await request(app)
          .delete(`/api/profiles/${follower.get('username')}/follow`)
          .set('Accept', 'application/json')
          .send();

        expect({body, status}).toMatchSnapshot();
      });
    });

    describe('with a logged in user', () => {
      describe('and an invalid username', () => {
        test('should return a 404', async () => {
          const {body, status} = await request(app)
            .delete(`/api/profiles/invalid/follow`)
            .set('Accept', 'application/json')
            .set('Authorization', `Token ${users.generateJWT(user)}`)
            .send();

          expect({body, status}).toMatchSnapshot();
        });
      });

      describe('and a valid username', () => {
        test('should return a 200', async () => {
          const {body, status} = await request(app)
            .delete(`/api/profiles/${user.get('username')}/follow`)
            .set('Accept', 'application/json')
            .set('Authorization', `Token ${users.generateJWT(follower)}`)
            .send();

          expect({body, status}).toMatchSnapshot();
        });
      });
    });
  });
});
