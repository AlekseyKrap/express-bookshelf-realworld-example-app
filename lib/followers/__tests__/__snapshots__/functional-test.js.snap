// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`follow DELETE /api/profiles/:username/follow with a logged in user and a valid username should return a 200 1`] = `
Object {
  "profile": Object {
    "bio": "Account Executive",
    "following": false,
    "image": "//www.gravatar.com/avatar/b21083770e1b6321bff13f8de9651d3d",
    "username": "wut",
  },
}
`;

exports[`follow DELETE /api/profiles/:username/follow with a logged in user and an invalid username should return a 404 1`] = `
Object {
  "errors": Object {
    "error": Object {
      "data": null,
      "isBoom": true,
      "isServer": false,
      "message": "Not Found",
      "output": Object {
        "headers": Object {},
        "payload": Object {
          "error": "Not Found",
          "message": "Not Found",
          "statusCode": 404,
        },
        "statusCode": 404,
      },
    },
    "message": "Not Found",
  },
}
`;

exports[`follow DELETE /api/profiles/:username/follow without a logged in user should return a 401 1`] = `
Object {
  "errors": Object {
    "error": Object {
      "code": "credentials_required",
      "data": Object {
        "UnauthorizedError": Array [
          "No authorization token was found",
        ],
      },
      "inner": Object {
        "message": "No authorization token was found",
      },
      "isBoom": true,
      "isServer": false,
      "message": "No authorization token was found",
      "name": "UnauthorizedError",
      "output": Object {
        "headers": Object {},
        "payload": Object {
          "error": "Unauthorized",
          "message": "No authorization token was found",
          "statusCode": 401,
        },
        "statusCode": 401,
      },
      "status": 401,
    },
    "message": "No authorization token was found",
  },
}
`;

exports[`follow POST /api/profiles/:username/follow with a logged in user and a valid username should return a 200 1`] = `
Object {
  "profile": Object {
    "bio": "Chief Medical Officer",
    "following": true,
    "image": "//www.gravatar.com/avatar/cb125e105a10574f985002573fc3886a",
    "username": "soh",
  },
}
`;

exports[`follow POST /api/profiles/:username/follow with a logged in user and an invalid username should return a 404 1`] = `
Object {
  "errors": Object {
    "error": Object {
      "data": null,
      "isBoom": true,
      "isServer": false,
      "message": "Not Found",
      "output": Object {
        "headers": Object {},
        "payload": Object {
          "error": "Not Found",
          "message": "Not Found",
          "statusCode": 404,
        },
        "statusCode": 404,
      },
    },
    "message": "Not Found",
  },
}
`;

exports[`follow POST /api/profiles/:username/follow without a logged in user should return a 401 1`] = `
Object {
  "errors": Object {
    "error": Object {
      "code": "credentials_required",
      "data": Object {
        "UnauthorizedError": Array [
          "No authorization token was found",
        ],
      },
      "inner": Object {
        "message": "No authorization token was found",
      },
      "isBoom": true,
      "isServer": false,
      "message": "No authorization token was found",
      "name": "UnauthorizedError",
      "output": Object {
        "headers": Object {},
        "payload": Object {
          "error": "Unauthorized",
          "message": "No authorization token was found",
          "statusCode": 401,
        },
        "statusCode": 401,
      },
      "status": 401,
    },
    "message": "No authorization token was found",
  },
}
`;
